### PR TITLE
Build .mjs files (ES modules) alongside CommonJS

### LINKED ISSUE #208: Publish ESM (.mjs) entry point to npm
While working on optimising the bundling of https://github.com/graphql/swapi-graphql — for _reasons_<sup>1</sup> — I discovered that it was `import`ing things from `graphql-relay-js`, but that the latter, unlike `graphql-js`, only had a CommonJS entry point published to npm.

It might be desirable to add an ES module build target to this package, modeled on what `graphql-js` is [already doing](https://github.com/graphql/graphql-js/blob/54f631ff3238affdddb4dc736fc282b620d70b15/package.json#L37).

---

<sup>1</sup> See https://twitter.com/motiz88/status/985110696668131328 - I'm investigating the [claim](https://github.com/apollographql/apollo-link/blob/a2025ad5a2eb7a1164d61a402e5deb227e14a84f/packages/apollo-link-schema/README.md) that schema execution on the front-end is impractical due to `graphql-js`'s [footprint](https://bundlephobia.com/result?p=graphql) in terms of bundle size. `swapi-graphql` simply happens to be a good, accessible testbed for just that (thanks @IvanGoncharov for the tip! 🙏)

### PR DESCRIPTION
**Pull request by motiz88**
Closes #208, with the actual build changes lifted nearly verbatim from `graphql-js` (e.g. https://github.com/graphql/graphql-js/pull/1244)

* Sets `sideEffects: false` _a la_ the recently-merged https://github.com/graphql/graphql-js/pull/1312.
* Works around an instance of https://github.com/webpack/webpack/issues/7032 (use of global `Buffer`) that would otherwise be encountered by users trying to actually bundle this as ESM with Webpack (e.g. what I'm currently doing with [`swapi-graphql`](https://github.com/graphql/swapi-graphql/compare/master...motiz88:webpack) as a testbed).
This is somewhat related to the issue with `process` fixed in https://github.com/graphql/graphql-js/commit/60c03ab23852074896fd9134c24c46ba595bf014 - except here, getting Node, Webpack and Flow to agree took some non-obvious fine-tuning, so I broke the solution out into its own heavily-commented file (`utils/buffer.js`). The Flow part can be simplified once ~https://github.com/facebook/flow/pull/6179~ hopefully lands.

UPDATE: https://github.com/facebook/flow/pull/6179 has landed and this branch incorporates the Flow 0.73 upgrade (which is also its own PR for good measure: https://github.com/graphql/graphql-relay-js/pull/210)

### CODE REVIEW

File: src/utils/base64.js [hunk 1]
```diff
@@ -7,6 +7,8 @@
  * @flow
  */
 
+import { Buffer } from './buffer';
+
```
Comment (IvanGoncharov) [inline src/utils/base64.js]: As I understand https://github.com/facebook/flow/issues/3723 was merged and will be available in the next release.
Does that mean you can just change this line to:
```js
import Buffer from 'buffer';
```
and everything will work?

File: src/utils/buffer.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import NodeBuffer from 'buffer';
+
+/**
+ * An alias for Node's Buffer class.
+ *
+ * Flow (tested with 0.69.0) knows about global.Buffer but not about
+ * require('buffer').Buffer.
+ * https://github.com/facebook/flow/issues/3723
+ */
```
Comment (IvanGoncharov) [inline src/utils/buffer.js]: @motiz88 Flow `0.72` released and it includes your fix 🎉 Can you please update this PR?

**Comment by motiz88** [comment]
@IvanGoncharov Done, thanks for pinging me.

@jaydenseric re: `.babelrc.js` and switching to `.mjs` source files, that all looks super valid, but I intentionally took my cue here from the way `graphql-js` currently does things, e.g. https://github.com/graphql/graphql-js/pull/1244. There's nothing stopping the maintainers of this package from changing the approach down the line if they deem it necessary.

File: src/__tests__/starWarsConnectionTests.js [hunk 3]
```diff
@@ -9,7 +9,7 @@
 
 import { expect } from 'chai';
 import { describe, it } from 'mocha';
-import { StarWarsSchema } from './starWarsSchema.js';
+import { StarWarsSchema } from './starWarsSchema';
```
Comment (IvanGoncharov) [inline src/__tests__/starWarsConnectionTests.js]: @motiz88 If you have time maybe it worth to split such changes into separate PR.
I think it would be easier to find someone from Facebook willing to merge such simple change
+ it will make the main PR smaller

**Comment by jaydenseric** [comment]
Guys, desperate for this. It seems there is nothing substantial blocking a merge?

I'm about to copy paste functions into my repo, write my own, or publish this PR as a disposable package to npm or something.

**Comment by WBad** [comment]
> Guys, desperate for this. It seems there is nothing substantial blocking a merge?
> 
> I'm about to copy paste functions into my repo, write my own, or publish this PR as a disposable package to npm or something.

@jaydenseric I did exactly that you can find the [package on npm](https://www.npmjs.com/package/@wbad/graphql-relay) the code on [gitlab](https://gitlab.com/wbad/graphql-relay-js)

**Comment by IvanGoncharov** [comment]
@motiz88 Sorry for the long delay.
We plan to merge this lib into `graphql-js` to provide the better maintenance and attention it deserves.
This should happen over the next few months and so closing this since `graphql-js` already provides mjs files.
